### PR TITLE
'sous artifact get' fixed and no longer requires -cluster flag + otpl-deploy smoke testing...

### PR DESCRIPTION
**Pull request by samsalisbury**
'sous artifact get' was crashing due to an injection fault, and now no longer needs the -cluster flag either. It was registered as 'sous artifact add get' but now it's just 'sous artifact get' This fix was incidental in writing tests for otpl-deploy integration, which also involved some tidy up of the smoke tests, and performance improvements.

This is a large PR and ends up touching some test infrastructure as well due to the growing number of tests requiring greater efficiency.. Smoke tests now run 3 minutes faster on my machine because old sous servers from passing tests are now stopped incrementally. Also when tests pass, all the configuration files, project files and logs are now deleted and replaced with a single empty file called "PASSED"... We could make this configurable in future if we care about transcripts from passing tests.

**UPDATE** these tests all pass locally for me but we have intermittent test server crashes in Travis, so I have added server crash detection and dumping of logs in those cases.

2018-07-24: Crash logs show that most server crashes are caused by failure to bind to port:
```
listen tcp 127.0.0.1:49326: bind: address already in use
```
This can be addressed by 2 methods: 1) grab ports later rather than all at boot time and 2) hold on to ports until they are needed by tests rather than racing over a longer time period. Making this change now.

### CODE REVIEW

**Comment by sous-machine-account** [comment]
# LGTM!

No errors, no warnings. :shipit:


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: test/smoke/bin.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+	}
+	return Bin{
+		BinPath:      path,
+		InstanceName: name,
+		BinName:      binName,
+		BaseDir:      baseDir,
+		Env:          map[string]string{},
+		ConfigDir:    filepath.Join(baseDir, "config"),
+		LogDir:       filepath.Join(baseDir, "logs"),
+		//Dir:       filepath.Join(baseDir, "work"),
```
Comment (JC1738) [inline test/smoke/bin.go]: not needed?

File: test/smoke/bin.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+// invocation is the invocation directly from the test, without any formatting
+// or manipulation.
+type invocation struct {
+	name, subcmd string
+	flags        Flags
+	args         []string
+}
+
+func (i invocation) String() string {
```
Comment (JC1738) [inline test/smoke/bin.go]: missing //String comment

File: test/smoke/matrixdef.go [hunk 3]
```diff
@@ -42,40 +49,61 @@ func (m MatrixDef) PrintDimensions() {
 	}
 }
 
-func (m *MatrixDef) AddDimension(name, desc string, values map[string]interface{}) {
+func (m *matrixDef) AddDimension(name, desc string, values map[string]interface{}) {
 	m.OrderedDimensionNames = append(m.OrderedDimensionNames, name)
 	m.OrderedDimensionDescs = append(m.OrderedDimensionDescs, desc)
 	m.Dimensions[name] = values
 }
```
Comment (JC1738) [inline test/smoke/matrixdef.go]: //FixedDimension comment

File: test/smoke/matrixdef.go [hunk 4]
```diff
@@ -42,40 +49,61 @@ func (m MatrixDef) PrintDimensions() {
...
 
+func (m matrixDef) FixedDimension(dimensionName, valueName string) matrixDef {
+	return m.Clone(func(dimension, value string) bool {
+		return dimension != dimensionName || value == valueName
+	})
+}
+
+func (m matrixDef) Clone(include func(dimension, value string) bool) matrixDef {
```
Comment (JC1738) [inline test/smoke/matrixdef.go]: //Clone comment

File: test/smoke/matrixdef.go [hunk 5]
```diff
@@ -42,40 +49,61 @@ func (m MatrixDef) PrintDimensions() {
...
 		m := c.Map()
 		fcfgs[i] = fixtureConfig{
 			Desc:      c.String(),
 			dbPrimary: m["store"].(bool),
-			projects:  m["project"].(ProjectList),
+			projects:  m["project"].(projectList),
 		}
 	}
 	return fcfgs
 }
```
Comment (JC1738) [inline test/smoke/matrixdef.go]: //Combinations

File: test/smoke/matrixdef.go [hunk 6]
```diff
@@ -87,47 +115,47 @@ func (m *MatrixDef) Combinations() []Combination {
...
+func concat(combos []combination) combination {
 	res := combos[0]
 	for _, c := range combos[1:] {
 		res = append(res, c...)
 	}
 	return res
 }
 
-func (c Combination) String() string {
+func (c combination) String() string {
```
Comment (JC1738) [inline test/smoke/matrixdef.go]: //String

File: test/smoke/matrixdef.go [hunk 7]
```diff
@@ -87,47 +115,47 @@ func (m *MatrixDef) Combinations() []Combination {
...
 	for _, p := range c {
 		names = append(names, p.Name)
 	}
 	return strings.Join(names, "/")
 }
 
-func (c Combination) Map() map[string]interface{} {
+func (c combination) Map() map[string]interface{} {
 	res := make(map[string]interface{}, len(c))
 	for _, p := range c {
```
Comment (JC1738) [inline test/smoke/matrixdef.go]: //Map

File: test/smoke/paralleltestfixture.go [hunk 8]
```diff
@@ -105,10 +84,10 @@ func (pf *ParallelTestFixture) recordTestStarted(t *testing.T) {
 // PTest is a test to run in parallel.
 type PTest struct {
 	Name string
-	Test func(*testing.T, *TestFixture)
+	Test func(*testing.T, *testFixture)
 }
 
-func (pf *ParallelTestFixture) RunMatrix(tests ...PTest) {
+func (pf *parallelTestFixture) RunMatrix(tests ...PTest) {
```
Comment (JC1738) [inline test/smoke/paralleltestfixture.go]: //RunMatrix

File: test/smoke/paralleltestfixture.go [hunk 9]
```diff
@@ -163,7 +143,7 @@ func (pf *ParallelTestFixture) recordTestStatus(t *testing.T) {
 	}
 }
 
-func (pfs *ParallelTestFixtureSet) PrintSummary() {
+func (pfs *parallelTestFixtureSet) PrintSummary() {
```
Comment (JC1738) [inline test/smoke/paralleltestfixture.go]: //Comments for rest of exported Functions

File: test/smoke/singularity.go [hunk 10]
```diff
@@ -16,17 +14,17 @@ import (
...
 	URL           string
 	client        *sing.Client
 	ClusterSuffix string
 }
 
-func NewSingularity(baseURL string) *Singularity {
-	return &Singularity{URL: baseURL, client: sing.NewClient(baseURL)}
+func newSingularity(baseURL string) *testSingularity {
+	return &testSingularity{URL: baseURL, client: sing.NewClient(baseURL)}
 }
```
Comment (JC1738) [inline test/smoke/singularity.go]: //Comments for exported functions in this file

File: test/smoke/sousflags.go [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+func (f *sousFlags) DeploymentIDFlags() *sousFlags {
+	if f == nil {
+		return nil
+	}
+	didFlags := f.ManifestIDFlags()
+	didFlags.cluster = f.cluster
+	return didFlags
+}
+
+func (f *sousFlags) SousDeployFlags() *sousFlags {
```
Comment (JC1738) [inline test/smoke/sousflags.go]: //Comment

File: test/smoke/testfixture.go [hunk 12]
```diff
@@ -89,29 +93,86 @@ func newTestFixture(t *testing.T, envDesc desc.EnvDesc, parent *ParallelTestFixt
 	return tf
 }
 
-func (f *TestFixture) DIDAndDefaultReqID(t *testing.T, repo, offset, flavor, cluster string) (sous.DeploymentID, string) {
+func (f *testFixture) Teardown(t *testing.T) {
```
Comment (JC1738) [inline test/smoke/testfixture.go]: //Comments for exported functions

File: test/smoke/testsousserver.go [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+	db, err := config.Database.DB()
+	if err != nil {
+		return err
+	}
+	mgr := storage.NewPostgresStateManager(db, logging.SilentLogSet())
+
+	return mgr.WriteState(state, sous.User{})
+}
+
+func (i *sousServer) Configure(config *config.Config, remoteGDMDir string, fcfg fixtureConfig) error {
```
Comment (JC1738) [inline test/smoke/testsousserver.go]: //Comment

**Comment by samsalisbury** [comment]
Thanks for review @JC1738, strictly speaking because I unexported all these types e.g. `combination`, `sousServer` their methods are also not exported (thus Hound did not make these comments, I believe it would have otherwise) did your local linter pick these up somehow?. These types were previously hidden behind the `smoke` build tag, but I removed that tag where possible, just leaving it on the tests themselves. In general though, comments are a good idea, I'll add them.

**Comment by JC1738** [comment]
@samsalisbury Got Jenkins to Pass, https://github.com/opentable/sous/pull/770/commits/7aa539e269ef2122b95cc06bee87e9e54bda9694 splitting up the runs in parallel for different parts of the smoke matrix.

**Comment by samsalisbury** [comment]
Oh that's great, thanks @JC1738 we should prob see about splitting these up in Travis as well for speed..

**Pull request merged**